### PR TITLE
chore: update datafusion to 30, arrow to 45

### PR DESCRIPTION
**Pull request by scsmithr**
# Description

Updates datafusion to 30, and arrow to 45.

# Related Issue(s)

N/A

# Documentation

N/A

### CODE REVIEW

File: rust/src/delta_datafusion.rs [hunk 1]
```diff
@@ -651,7 +651,7 @@ pub(crate) fn partitioned_file_from_action(
 
     let ts_secs = action.modification_time / 1000;
     let ts_ns = (action.modification_time % 1000) * 1_000_000;
-    let last_modified = DateTime::<Utc>::from_utc(
+    let last_modified = DateTime::<Utc>::from_naive_utc_and_offset(
```
Comment (scsmithr) [inline rust/src/delta_datafusion.rs:654]: Removes deprecation warning.

File: rust/src/operations/optimize.rs [hunk 2]
```diff
@@ -1180,11 +1180,11 @@ pub(super) mod zorder {
             .map(|i| (i * value_size) as i32)
             .collect::<Vec<i32>>();
 
-        let out_arr = BinaryArray::new_unchecked(
+        let out_arr = BinaryArray::try_new(
```
Comment (scsmithr) [inline rust/src/operations/optimize.rs:1183]: `new_unchecked` was made unsafe. This does the same thing, just with some validation checks. I'm not aware of any significant performance hits with this extra check.

File: Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 [workspace]
 members = ["rust", "python"]
 exclude = ["proofs", "delta-inspect"]
+resolver = "2"
```
Comment (scsmithr) [inline Cargo.toml:4]: Rust 1.72 emits a warning about not specifying this now. "2" is the default.

**Comment by wjones127** [comment]
Would you be willing to fix that Python test? should just need to replace the expected output.

https://github.com/delta-io/delta-rs/actions/runs/6015750742/job/16318366781?pr=1606#step:9:168

**Comment by scsmithr** [comment]
> Would you be willing to fix that Python test? should just need to replace the expected output.
> 
> https://github.com/delta-io/delta-rs/actions/runs/6015750742/job/16318366781?pr=1606#step:9:168

Updated in latest commit.

**Comment by scsmithr** [comment]
> Would you mind rebasing this branch @scsmithr ? I just merged @wjones127 changes which simplified some changes and removed some of the deprecation warnings

Done

**Review by rtyler — APPROVED** [review]

**Pull request merged**